src/cryptsetup/cryptsetup-tpm2.c

```diff
@@ -64,7 +114,53 @@ int acquire_tpm2_key(
                 blob = loaded_blob;
         }
 
-        return tpm2_unseal(device, pcr_mask, pcr_bank, primary_alg, blob, blob_size, policy_hash, policy_hash_size, NULL, ret_decrypted_key, ret_decrypted_key_size);
+        if (flags & TPM2_FLAGS_USE_PIN) {
+                for (int i = 5;; i--) {
+                        _cleanup_(erase_and_freep) char *pin_str = NULL;
+
+                        if (i <= 0)
+                                return -EACCES;
+
+                        r = get_pin(until, ask_password_flags, headless, &pin_str);
+                        if (r == -EPERM)
+                                continue;
+                        if (r < 0)
+                                return r;
+
+                        r = tpm2_unseal(
+                                        device,
+                                        pcr_mask,
+                                        pcr_bank,
+                                        primary_alg,
+                                        blob,
+                                        blob_size,
+                                        policy_hash,
+                                        policy_hash_size,
+                                        pin_str,
+                                        ret_decrypted_key,
+                                        ret_decrypted_key_size);
+                        /* We get this error in case there is an authentication policy mismatch. This should
+                         * not happen, but this avoids confusing behavior, just in case. */
+                        if (IN_SET(r, -EPERM, -ENOLCK))
+                                return r;
+                        if (r < 0)
+                                continue;
+
+                        return r;
+                }
+        } else
+                return tpm2_unseal(
```

Comment:
no need for an else here - to reduce indentation, invert the check and return tpm2_unseal inside the if, and have the for block after